File: Cloud Storage System/Level 3/Python/user_repository.py
# user_repository.py

from user import User

class UserRepository:
    def __init__(self):
        self.users = {}  # user_id -> User

    def add_user(self, user_id: str, capacity: int) -> bool:
        if user_id in self.users:
            return False
        self.users[user_id] = User(user_id, capacity)
        return True

    def get_user(self, user_id: str) -> User | None:
        return self.users.get(user_id)

    def delete_user(self, user_id: str):
        if user_id in self.users:
            del self.users[user_id]

    def merge_users(self, user1_id: str, user2_id: str) -> int | None:
        if user1_id == "admin" or user2_id == "admin":
            return None
        if user1_id not in self.users or user2_id not in self.users:
            return None

        user1 = self.users[user1_id]
        user2 = self.users[user2_id]

        for file_path in user2.files:
            user1.files.add(file_path)

        user1.used += user2.used
        user1.capacity += user2.capacity

        self.delete_user(user2_id)
        return user1.get_remaining_capacity()

File: Cloud Storage System/Level 3/C#/user_repository.cs
// UserRepository.cs
using System.Collections.Generic;

public class UserRepository
{
    private readonly Dictionary<string, User> _users = new();

    public bool AddUser(string userId, int capacity)
    {
        if (_users.ContainsKey(userId)) return false;
        _users[userId] = new User(userId, capacity);
        return true;
    }

    public User? GetUser(string userId)
    {
        return _users.TryGetValue(userId, out var user) ? user : null;
    }

    public void DeleteUser(string userId)
    {
        _users.Remove(userId);
    }

    public int? MergeUsers(string user1Id, string user2Id)
    {
        if (user1Id == "admin" || user2Id == "admin") return null;
        if (!_users.ContainsKey(user1Id) || !_users.ContainsKey(user2Id)) return null;

        var user1 = _users[user1Id];
        var user2 = _users[user2Id];

        user1.MergeFrom(user2);
        DeleteUser(user2Id);

        return user1.RemainingCapacity();
    }
}
